app/models/topic_user.rb

```diff
@@ -131,6 +131,9 @@ def change(user_id, topic_id, attrs)
 
         if rows == 0
           create_missing_record(user_id, topic_id, attrs)
+        elsif topic_user = TopicUser.find_by(topic_id: topic_id, user_id: user_id, notifications_reason_id: nil, notifications_changed_at: nil)
```

Comment:
I am not sure here, this is changing 

update
- if missing create

to 

update
- if existing ... lookup
- then update again

This is an extra query... I would prefer to keep the usual use case here as single query.

Reply:
This is a tricky part. TopicUser record are created in two cases. When the user visits the topic for the first time or when post which was already seen by the user is moved to another topic (then we create topic_user with notification level set as regular)

So logic we have here is
- If there is no topic_user record, create one and while you are creating, check if the user wanted to automatically and immediately track topic 

- if there is already topic_user, without notifications_reason_id (so it was created when the post was moved and notification level was not changed by the user) - check if the user wanted to automatically track topic and do that for them. 

I was thinking as well if we should set correct notification level when a post is moved and we create that topic_user. However, if the user was tracking the original topic, it doesn't mean that they want to track automatically new topic to which post was moved to

I will try to solve without adding a new query